File: packages/protocols/x402-ampersend-sdk/python/ampersend-sdk/src/ampersend_sdk/ampersend/treasurer.py
import logging
import uuid
from typing import Any

from x402_a2a.types import (
    PaymentStatus,
    x402PaymentRequiredResponse,
)

from ampersend_sdk.x402 import X402Authorization, X402Treasurer, X402Wallet

from .client import ApiClient
from .types import (
    PaymentEvent,
    PaymentEventAccepted,
    PaymentEventError,
    PaymentEventRejected,
    PaymentEventSending,
)

logger = logging.getLogger(__name__)


class AmpersendTreasurer(X402Treasurer):
    """
    Ampersend API treasurer with authorization checks and event reporting.

    Works with both EOA and smart account payment methods via the
    X402Wallet protocol.
    """

    def __init__(self, api_client: ApiClient, wallet: X402Wallet):
        """
        Initialize Ampersend treasurer.

        Args:
            api_client: ApiClient instance for authorization checks
            wallet: X402Wallet for creating payment payloads
        """
        self._api_client = api_client
        self._wallet = wallet

    async def onPaymentRequired(
        self,
        payment_required: x402PaymentRequiredResponse,
        context: dict[str, Any] | None = None,
    ) -> X402Authorization | None:
        result = await self._api_client.authorize_payment(
            payment_required.accepts, context
        )

        # Check if any requirements were authorized
        if not result.authorized.requirements:
            # Log rejection reasons for debugging
            reasons = ", ".join(
                [f"{r.requirement.resource}: {r.reason}" for r in result.rejected]
            )
            logger.info(
                "No requirements authorized. Reasons: %s",
                reasons,
            )
            return None

        # Use recommended requirement (or first if recommended is None)
        recommended_index = (
            result.authorized.recommended
            if result.authorized.recommended is not None
            else 0
        )

        # Validate recommended index is within bounds
        if recommended_index >= len(result.authorized.requirements):
            raise ValueError(
                f"Invalid recommended index {recommended_index}, "
                f"only {len(result.authorized.requirements)} requirements authorized"
            )

        authorized_req = result.authorized.requirements[recommended_index]

        # Create payment with wallet using the authorized requirement
        payment = self._wallet.create_payment(
            requirements=authorized_req.requirement,
        )
        authorization_id = uuid.uuid4().hex

        await self._api_client.report_payment_event(
            event_id=authorization_id,
            payment=payment,
            event=PaymentEventSending(),
        )

        return X402Authorization(
            authorization_id=authorization_id,
            payment=payment,
            selected_requirement=authorized_req.requirement,
        )

    async def onStatus(
        self,
        status: PaymentStatus,
        authorization: X402Authorization,
        context: dict[str, Any] | None = None,
    ) -> None:
        # Map status to appropriate event type
        event: PaymentEvent | None = None

        if status == PaymentStatus.PAYMENT_SUBMITTED:
            event = PaymentEventSending()
        elif status == PaymentStatus.PAYMENT_FAILED:
            # Extract reason from context if available, otherwise use default
            reason = "Payment processing failed"
            if context and isinstance(context, dict):
                reason = context.get("reason", reason)
            event = PaymentEventError(reason=reason)
        elif status == PaymentStatus.PAYMENT_REJECTED:
            # Extract reason from context if available, otherwise use default
            reason = "Payment rejected by server"
            if context and isinstance(context, dict):
                reason = context.get("reason", reason)
            event = PaymentEventRejected(reason=reason)
        elif status in (
            PaymentStatus.PAYMENT_VERIFIED,
            PaymentStatus.PAYMENT_COMPLETED,
        ):
            event = PaymentEventAccepted()

        # Only report events we care about
        if event is None:
            return

        await self._api_client.report_payment_event(
            event_id=authorization.authorization_id,
            payment=authorization.payment,
            event=event,
        )


DEFAULT_API_URL = "https://api.ampersend.ai"


def create_ampersend_treasurer(
    *,
    smart_account_address: str,
    session_key_private_key: str,
    api_url: str = DEFAULT_API_URL,
) -> AmpersendTreasurer:
    """
    Create an AmpersendTreasurer with minimal configuration.

    This is the recommended way to create a treasurer for most use cases.

    Args:
        smart_account_address: The smart account address (0x...)
        session_key_private_key: The session key private key (0x...)
        api_url: Ampersend API URL (defaults to production)

    Returns:
        Configured AmpersendTreasurer ready for use

    Example:
        treasurer = create_ampersend_treasurer(
            smart_account_address="0x1234...",
            session_key_private_key="0xabcd...",
        )
    """
    from ampersend_sdk.smart_account import SmartAccountConfig
    from ampersend_sdk.x402.wallets.smart_account import SmartAccountWallet

    from .types import ApiClientOptions

    wallet = SmartAccountWallet(
        config=SmartAccountConfig(
            session_key=session_key_private_key,
            smart_account_address=smart_account_address,
        )
    )

    api_client = ApiClient(
        options=ApiClientOptions(
            base_url=api_url,
            session_key_private_key=session_key_private_key,
            agent_address=smart_account_address,
        )
    )

    return AmpersendTreasurer(api_client=api_client, wallet=wallet)

File: packages/protocols/x402-ampersend-sdk/typescript/packages/ampersend-sdk/src/ampersend/treasurer.ts
import type { Address, Hex } from "viem"
import { privateKeyToAccount } from "viem/accounts"

import type { PaymentEvent } from "../mcp/client/index.ts"
import { OWNABLE_VALIDATOR } from "../smart-account/index.ts"
import {
  createWalletFromConfig,
  type Authorization,
  type PaymentContext,
  type PaymentStatus,
  type SmartAccountWalletConfig,
  type WalletConfig,
  type X402Treasurer,
  type X402Wallet,
} from "../x402/index.ts"
import { ApiClient } from "./client.ts"

/** Default Ampersend API URL */
const DEFAULT_API_URL = "https://api.ampersend.ai"

/** Default chain ID (Base mainnet) */
const DEFAULT_CHAIN_ID = 8453

/**
 * Simplified configuration for quick setup with smart accounts.
 * This is the recommended way to configure the treasurer for most use cases.
 *
 * @example
 * ```typescript
 * const treasurer = createAmpersendTreasurer({
 *   smartAccountAddress: "0x...",
 *   sessionKeyPrivateKey: "0x...",
 * })
 * ```
 */
export interface SimpleAmpersendTreasurerConfig {
  /** Smart account address */
  smartAccountAddress: Address
  /** Session key private key for signing */
  sessionKeyPrivateKey: Hex
  /** Ampersend API URL (defaults to production) */
  apiUrl?: string
  /** Chain ID (defaults to Base mainnet 8453) */
  chainId?: number
}

/**
 * Full configuration for advanced use cases with complete wallet control.
 * Use this when you need EOA wallets or custom authentication settings.
 */
export interface FullAmpersendTreasurerConfig {
  /** Base URL of the Ampersend API server */
  apiUrl: string
  /** Wallet configuration (EOA or Smart Account) */
  walletConfig: WalletConfig
  /** Optional authentication configuration */
  authConfig?: {
    /** SIWE domain for authentication */
    domain?: string
    /** SIWE statement for authentication */
    statement?: string
  }
}

/**
 * Configuration for the Ampersend treasurer.
 * Can be either simplified (recommended) or full configuration.
 */
export type AmpersendTreasurerConfig = SimpleAmpersendTreasurerConfig | FullAmpersendTreasurerConfig

/**
 * Type guard to check if config is the simplified format
 */
function isSimpleConfig(config: AmpersendTreasurerConfig): config is SimpleAmpersendTreasurerConfig {
  return "smartAccountAddress" in config && "sessionKeyPrivateKey" in config && !("walletConfig" in config)
}

/**
 * AmpersendTreasurer - Ampersend API-based payment authorization with X402Treasurer pattern
 *
 * This treasurer:
 * 1. Authenticates with the Ampersend API using SIWE
 * 2. Requests payment authorization from the API before creating payments
 * 3. Creates payments only when authorized by the API
 * 4. Reports payment lifecycle events back to the API for tracking
 *
 * @example
 * ```typescript
 * const treasurer = createAmpersendTreasurer({
 *   apiUrl: "https://api.example.com",
 *   walletConfig: { type: "eoa", privateKey: "0x..." }
 * })
 * await initializeProxyServer({ transport, treasurer })
 * ```
 */
export class AmpersendTreasurer implements X402Treasurer {
  constructor(
    private apiClient: ApiClient,
    private wallet: X402Wallet,
  ) {}

  /**
   * Requests payment authorization from API before creating payment.
   * Only creates payment if API authorizes it.
   */
  async onPaymentRequired(requirements: Array<any>, context?: PaymentContext): Promise<Authorization | null> {
    try {
      // Authorize payment with API
      const response = await this.apiClient.authorizePayment(requirements as any, context)

      // Check if any requirements were authorized
      if (response.authorized.requirements.length === 0) {
        // Log rejection reasons for debugging
        const reasons = response.rejected.map((r) => `${r.requirement.resource}: ${r.reason}`).join(", ")
        console.log(`[AmpersendTreasurer] No requirements authorized. Reasons: ${reasons || "None provided"}`)
        return null // Decline
      }

      // Use recommended requirement (or first if recommended is null)
      const recommendedIndex = response.authorized.recommended ?? 0
      const authorizedReq = response.authorized.requirements[recommendedIndex]

      if (!authorizedReq) {
        throw new Error("Recommended requirement index out of bounds")
      }

      // Create payment with wallet using the authorized requirement
      // Note: Type assertion needed because ampersend PaymentRequirements uses string for network,
      // while x402 PaymentRequirements uses specific network literals. Runtime compatible.
      const payment = await this.wallet.createPayment(authorizedReq.requirement as any)

      return {
        payment,
        authorizationId: crypto.randomUUID(),
      }
    } catch (error) {
      console.error("[AmpersendTreasurer] Payment authorization failed:", error)
      return null
    }
  }

  /**
   * Reports payment status updates back to API for tracking.
   * Logs errors but doesn't fail on tracking errors.
   */
  async onStatus(status: PaymentStatus, authorization: Authorization, _context?: PaymentContext): Promise<void> {
    try {
      // Map status to event type for API
      const event = this.mapStatusToEvent(status)
      await this.apiClient.reportPaymentEvent(authorization.authorizationId, authorization.payment, event)
    } catch (error) {
      // Log but don't fail on event tracking errors
      console.error(`[AmpersendTreasurer] Failed to report status ${status}:`, error)
    }
  }

  /**
   * Maps X402 PaymentStatus to legacy PaymentEvent for API compatibility
   */
  private mapStatusToEvent(status: PaymentStatus): PaymentEvent {
    switch (status) {
      case "sending":
        return { type: "sending" }
      case "accepted":
        return { type: "accepted" }
      case "rejected":
        return { type: "rejected", reason: "Payment rejected by server" }
      case "declined":
        return { type: "rejected", reason: "Payment declined by treasurer" }
      case "error":
        return { type: "error", reason: "Payment processing error" }
    }
  }
}

/**
 * Creates an Ampersend treasurer that consults the Ampersend API before making payments.
 *
 * This treasurer:
 * 1. Authenticates with the Ampersend API using SIWE
 * 2. Requests payment authorization from the API
 * 3. Creates payments only when authorized
 * 4. Reports payment lifecycle events back to the API
 *
 * @example Simple setup (recommended):
 * ```typescript
 * const treasurer = createAmpersendTreasurer({
 *   smartAccountAddress: "0x...",
 *   sessionKeyPrivateKey: "0x...",
 * })
 * ```
 *
 * @example Full control:
 * ```typescript
 * const treasurer = createAmpersendTreasurer({
 *   apiUrl: "https://api.ampersend.ai",
 *   walletConfig: { type: "eoa", privateKey: "0x..." }
 * })
 * ```
 *
 * @param config - Configuration for the Ampersend treasurer
 * @returns An X402Treasurer implementation
 */
export function createAmpersendTreasurer(config: AmpersendTreasurerConfig): X402Treasurer {
  if (isSimpleConfig(config)) {
    // Simple config - build wallet config automatically
    const walletConfig: SmartAccountWalletConfig = {
      type: "smart-account",
      smartAccountAddress: config.smartAccountAddress,
      sessionKeyPrivateKey: config.sessionKeyPrivateKey,
      chainId: config.chainId ?? DEFAULT_CHAIN_ID,
      validatorAddress: OWNABLE_VALIDATOR,
    }

    const apiClient = new ApiClient({
      baseUrl: config.apiUrl ?? DEFAULT_API_URL,
      sessionKeyPrivateKey: config.sessionKeyPrivateKey,
      agentAddress: config.smartAccountAddress,
      timeout: 30000,
    })

    const wallet = createWalletFromConfig(walletConfig)
    return new AmpersendTreasurer(apiClient, wallet)
  }

  // Full config - existing behavior
  const { apiUrl, authConfig, walletConfig } = config

  // Determine which private key to use for API authentication
  const authPrivateKey = walletConfig.type === "eoa" ? walletConfig.privateKey : walletConfig.sessionKeyPrivateKey

  // Derive agent address from wallet config
  const agentAddress =
    walletConfig.type === "smart-account"
      ? walletConfig.smartAccountAddress
      : privateKeyToAccount(walletConfig.privateKey).address

  // Create API client
  const apiClient = new ApiClient({
    baseUrl: apiUrl,
    sessionKeyPrivateKey: authPrivateKey,
    agentAddress,
    timeout: 30000,
    ...authConfig,
  })

  // Create wallet from configuration
  const wallet = createWalletFromConfig(walletConfig)

  return new AmpersendTreasurer(apiClient, wallet)
}
